From 42879b01f8f8a2208dccdf155b417df18bd35cae Mon Sep 17 00:00:00 2001
From: Rob Clark <rclark@redhat.com>
Date: Wed, 17 Dec 2014 21:38:43 -0500
Subject: [drm] vmwgfx: Fix a potential infinite spin waiting for fifo idle

Message-id: <1418852380-13061-33-git-send-email-rclark@redhat.com>
Patchwork-id: 102163
O-Subject: [RHEL7 drm 32/89] drm/vmwgfx: Fix a potential infinite spin waiting for fifo idle
Bugzilla: 1173317
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Dave Airlie <airlied@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1173317
Upstream: since v3.16.4

commit e7d2e5620110eeb6e618bb23b7dc5aad0b7028b4

Author:     Thomas Hellstrom <thellstrom@vmware.com>
AuthorDate: Thu Aug 28 11:53:23 2014 +0200
Commit:     Rob Clark <rclark@redhat.com>
CommitDate: Wed Dec 10 17:16:30 2014 -0500

    drm/vmwgfx: Fix a potential infinite spin waiting for fifo idle

    commit f01ea0c3d9db536c64d47922716d8b3b8f21d850 upstream.

    The code waiting for fifo idle was incorrect and could possibly spin
    forever under certain circumstances.

    Signed-off-by: Thomas Hellstrom <thellstrom@vmware.com>
    Reported-by: Mark Sheldon <markshel@vmware.com>
    Reviewed-by: Jakob Bornecrantz <jakob@vmware.com>
    Reivewed-by: Mark Sheldon <markshel@vmware.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Signed-off-by: Rob Clark <rclark@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/vmwgfx/vmwgfx_fifo.c b/drivers/gpu/drm/vmwgfx/vmwgfx_fifo.c
index 6ccd993..6eae14d 100644
--- a/drivers/gpu/drm/vmwgfx/vmwgfx_fifo.c
+++ b/drivers/gpu/drm/vmwgfx/vmwgfx_fifo.c
@@ -180,8 +180,9 @@ void vmw_fifo_release(struct vmw_private *dev_priv, struct vmw_fifo_state *fifo)
 
  mutex_lock(&dev_priv->hw_mutex);
 
+ vmw_write(dev_priv, SVGA_REG_SYNC, SVGA_SYNC_GENERIC);
  while (vmw_read(dev_priv, SVGA_REG_BUSY) != 0)
-  vmw_write(dev_priv, SVGA_REG_SYNC, SVGA_SYNC_GENERIC);
+  ;
 
  dev_priv->last_read_seqno = ioread32(fifo_mem + SVGA_FIFO_FENCE);
 
-- 
1.7.1